src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp

```diff
@@ -657,11 +657,6 @@ void ShenandoahGeneration::establish_usage(size_t num_regions, size_t num_bytes,
 
 void ShenandoahGeneration::increase_used(size_t bytes) {
   Atomic::add(&_used, bytes);
-  // This detects arithmetic wraparound on _used.  Non-generational mode does not keep track of _affiliated_region_count
-  // TODO: REMOVE IS_GLOBAL() QUALIFIER AFTER WE FIX GLOBAL AFFILIATED REGION ACCOUNTING
-  assert(is_global() || ShenandoahHeap::heap()->is_full_gc_in_progress() ||
-         (_used + _humongous_waste <= _affiliated_region_count * ShenandoahHeapRegion::region_size_bytes()),
-         "used cannot exceed regions");
 }
```

Comment:
Note that C++ subexpression evaluation order is undefined.  Here is an example of what can go wrong with the removed assertion:

ThisThread:  fetches _affiliated_region_count * begins to multiply with region_size
OtherThread: increases _used and increases _affiliated_region_count appropriately (for a large allocation)
ThisThread: fetches _used and observes assert violation

Reply:
We have seen this race manifest in actual testing, which is what has motivated us to remove the assertions.